docs/static/persistent-queues.asciidoc

```diff
@@ -1,44 +1,228 @@
+
+[[persistent-queues-benefits]]
+==== Benefits of persistent queues
+
+A persistent queue (PQ):
+
+* Helps protect against message loss during a normal shutdown and when Logstash
+is terminated abnormally. 
+If Logstash is restarted while events are in-flight,
+Logstash attempts to deliver messages stored in the persistent queue until
+delivery succeeds at least once.  
+* Can absorb bursts of events without needing an external buffering mechanism like Redis or Apache Kafka.
+* Can be useful for Logstash deployments that need large buffers.
+You can enable persistent queues to buffer events on disk in a buffered publish-subscriber model without a message broker such as Redis, RabbitMQ, or
+Apache Kafka). 
+
+NOTE: Persistent queues are disabled by default.
+To enable them, check out <<configuring-persistent-queues>>.
 
 [[persistent-queues-limitations]]
-==== Limitations of Persistent Queues
+==== Limitations of persistent queues
+
+Persistent queues do not solve these problems:
+
+* Input plugins that do not use a request-response protocol cannot be protected from data loss. Tcp, udp, zeromq push+pull, and many other inputs do not have a mechanism to acknowledge receipt to the sender. 
+(Plugins such as beats and http, which *do* have an acknowledgement capability, are well protected by this queue.)
+* Data may be lost if an abnormal shutdown occurs before the checkpoint file has been committed.
+* A persistent queue does not handle permanent machine failures such as disk corruption, disk failure, and machine loss. 
+The data persisted to disk is not replicated.
+
+
+[[configuring-persistent-queues]]
+==== Configuring persistent queues
+
+To configure persistent queues, specify options in the Logstash <<logstash-settings-file,settings file>>:
+
+`queue.type`:: Specify `persisted` to enable persistent queues. By default, persistent queues are disabled (default: `queue.type: memory`).
+`path.queue`:: The directory path where the data files will be stored. By default, the files are stored in `path.data/queue`.
+`queue.type`:: Specify `persisted` to enable persistent queues. By default, persistent queues are disabled (default: `queue.type: memory`).
+`path.queue`:: The directory path where the data files will be stored. By default, the files are stored in `path.data/queue`. 
+`queue.page_capacity`:: The maximum size of a queue page in bytes. The queue data consists of append-only files called "pages". The default size is 64mb. Changing this value is unlikely to have performance benefits.
+`queue.drain`:: Specify `true` if you want Logstash to wait until the persistent queue is drained before shutting down. The amount of time it takes to drain the queue depends on the number of events that have accumulated in the queue. Therefore, you should avoid using this setting unless the queue, even when full, is relatively small and can be drained quickly. 
```

Comment:
I'd like to add a little detail here as to why you might want to use this, but I think this can wait for another PR

Reply:
Upgrading (6.x timeframe) is the scenario I know about. What others? 
I have a section devoted to draining the queue (starting at line 215). That's probably a better place for this info.